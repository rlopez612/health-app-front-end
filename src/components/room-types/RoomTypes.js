import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import style from './Rooms.module.css';
import Button from '../button/Button';
import Card from '../card/Card';
import Modal from '../modal/Modal';
import Spinner from '../spinner/Spinner';
import HttpHelper from '../../utils/HttpHelper';

/**
 * @name RoomTypes
 * @description Room-types page
 * @return component
 */
const RoomTypes = () => {
  const [apiError, setApiError] = useState(false);
  const [loading, setLoading] = useState(false);
  const [rooms, setRooms] = useState([]);

  useEffect(() => {
    setLoading(true);
    HttpHelper('/room-types', 'GET')
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw new Error('Something went wrong');
      })
      .then((data) => {
        setLoading(false);
        setRooms(data);
      })
      .catch(() => {
        setLoading(false);
        setApiError(true);
      });
  }, []);

  /**
 * @name createRoomTypeDisplays
 * @description creates card to display each room-type
 * @return component[]
 */
  const createRoomTypeDisplays = () => rooms.map((room) => (
    <Card key={room.id}>
      <p className={style.Text}>
        <strong>Room Type: </strong>
        {room.name}
      </p>
      <p className={style.Text}>
        <strong>Description: </strong>
        {room.description}
      </p>
      <p className={style.Text}>
        <strong>Rate: </strong>
        {room.rate}
      </p>
      <p className={style.Text}>
        <strong>Status: </strong>
        {room.active ? 'Active' : 'Inactive'}
      </p>

      <div className={style.Container}>
        <Link className={style.Link} to={`/room-types/edit/${room.id}`}><Button color="Primary" type="button">Edit</Button></Link>
      </div>
    </Card>
  ));

  return (
    <div className={style.container}>
      {loading && <Spinner />}
      {apiError && <Modal message="Oops something went wrong" reset={() => setApiError(false)} />}

      <div className={style.Header}>
        <h1>Room Types</h1>
        <Link to="/room-types/create"><Button color="Primary" type="button">Create</Button></Link>
      </div>

      <div className={style.flex}>
        {createRoomTypeDisplays()}
      </div>

    </div>
  );
};

export default RoomTypes;
