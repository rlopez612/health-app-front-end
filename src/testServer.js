import { rest } from 'msw';
import { setupServer } from 'msw/node';

// Establish mock server for all endpoints with 'happy path' mock data
const server = setupServer(
  
  rest.get('http://localhost:8080/room-types', (req, res, ctx) => (
    res(ctx.json([
      {
        id: 1,
        name: 'Executive Suite',
        description: 'Full business suite with office',
        rate: 99.99,
        active: true
      },
      {
        id: 2,
        name: 'King Double',
        description: 'Double king non-smoking',
        rate: 149.99,
        active: true
      }
    ]))
  )),

  rest.get('http://localhost:8080/room-types/:id', (req, res, ctx) => {
    const { id } = req.params;
    return res(
      ctx.json({
        id,
        name: 'Executive Suite',
        description: 'Full business suite with office',
        rate: 99.99,
        active: true
      })
    );
  }),

  rest.post('http://localhost:8080/room-types', (req, res, ctx) => res(ctx.json()))
);

// Establish requests interception layer before all tests.
beforeAll(() => server.listen());
// Reset handlers back to the server defaults above when overriding for sad path tests
afterEach(() => server.resetHandlers());
// Clean up after all tests are done, closing the mock server
afterAll(() => server.close());

export { rest, server };
