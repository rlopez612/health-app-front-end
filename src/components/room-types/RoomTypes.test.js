import React from 'react';
import {
  render, screen, waitFor, fireEvent
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { BrowserRouter, Router } from 'react-router-dom';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import { rest, server } from '../../testServer';
import RoomTypes from './RoomTypes';

it('Renders the RoomTypes page', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <RoomTypes user={{ role: 'manager' }} />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Fetches data from the API', async () => {
  render(<RoomTypes user={{ role: 'manager' }} />, { wrapper: BrowserRouter });
  const heading = screen.getByRole('heading');
  const roomDescription = await waitFor(() => screen.findByText('Double king non-smoking'));

  expect(heading).toHaveTextContent('Room Types');
  expect(roomDescription).toBeVisible();
});

it('Should display error message when there is a server error', async () => {
  // Override default server response from testServer.js to test 'sad path'
  server.use(
    rest.get('http://localhost:8080/room-types', (req, res, ctx) => res(ctx.status(500)))
  );

  render(<RoomTypes user={{ role: 'manager' }} />, { wrapper: BrowserRouter });
  const errorMsg = await waitFor(() => screen.getByText('Oops something went wrong'));

  expect(errorMsg).toBeVisible();
  // Click button to close modal and remove error message
  fireEvent.click(screen.getByRole('button', { name: /ok/i }));
  expect(screen.queryByText(errorMsg)).toBeNull();
});

it('should redirect to the create room type page when clicking on the create button', () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <RoomTypes user={{ role: 'manager' }} />
    </Router>
  );
  const button = screen.getByText('Create');
  fireEvent.click(button);

  expect(history.location.pathname).toBe('/room-types/create');
});

it('should redirect to edit room type page when clicking on the edit button', async () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <RoomTypes user={{ role: 'manager' }} />
    </Router>
  );
  const buttons = await screen.findAllByRole('button', { name: /edit/i });
  fireEvent.click(buttons[0]);

  expect(buttons).toHaveLength(2);
  expect(history.location.pathname).toBe('/room-types/edit/1');
});
