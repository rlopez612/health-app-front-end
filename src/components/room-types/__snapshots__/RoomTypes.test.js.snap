// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the RoomTypes page 1`] = `
<div
  className="container"
>
  <div
    className="Header"
  >
    <h1>
      Room Types
    </h1>
    <a
      href="/room-types/create"
      onClick={[Function]}
    >
      <button
        className="Button Primary"
        type="button"
      >
        Create
      </button>
    </a>
  </div>
  <div
    className="grid"
  />
</div>
`;
